Open Problem: Human Readable Naming (#395)
ipfs/notes: OPEN_PROBLEMS/HUMAN_READABLE_NAMING.md

## Short Description
> In one sentence or paragraph.

> You can only have two of three properties for a name: Human-meaningful, Secure and/or Decentralized. This is [Zooko's Trilemma](https://en.wikipedia.org/wiki/Zooko%27s_triangle). Can we have all 3, or even more? Can context related to some data help solve this problem?

Reviewer: Doesn't the very page we're linking to state that we can, indeed, have all 3 and more?

Serious question as it's unclear if we're just using the triangle as a background introduction or taking it seriously, the latter meaning we are restricting the scope to exclude said existing solutions.

Reviewer: Here's the issue I have with Zooko's triangle: people often equate it to the CAP theorem, overlooking the fact that one is a proven theorem and the other is a poorly formalised conjecture in a blog post.

My understanding is that there are, in fact, solutions that achieve all three properties, _depending_ on how you define those properties ("The amount of damage a malicious entity can inflict on the system should be as low as possible." is rather meaningless). Per the Wikipedia page, both Swartz' and Szabo's papers achieve security under Byzantine assumptions. The question then becomes whether that's _enough_.

Author: @yiannisbot's review is correct. @jsoares would you like to add a clarification on the long description that the goal is to achieve the whole 3 and at scale?

Reviewer: Your answer did not shed any light for me. Do the aforementioned solutions fail to achieve all three properties and, if so, why? Is it just a scalability issue?

While I'll admit this is also a pet peeve of mine, I'm not being pedantic: if we turn this into an RFP, the solution should not be something linked on a Wikipedia page.

## Long Description

Content naming is an integral part of any content-oriented network architecture. In contrast to today’s approach of securing the channel and authenticating the end-host when receiving some content, in content-centric networks, it is the content itself that has to be secured and authenticated. Doing so enables the very important property of location-independent caching of content - that is, clients don’t need to trust the machine that provides the content, as long as one can verify that the content is what it claims to be. The claim is verified through the content name.

That said, the naming scheme needs to integrate some properties to guarantee security and some others to enable non-security-related functionality.

There are generally three types of names that have so far been investigated and implemented in the area of content-oriented networks: flat, normally hash-based names (as seen in IPFS and [NetInf](https://core.ac.uk/download/pdf/11433344.pdf)), hierarchical, as seen in [NDN](https://named-data.net) and [tag-based](http://conferences2.sigcomm.org/acm-icn/2014/papers/p17.pdf).

> We start by dividing between two types of properties for content names, that is, those that a name “has to have” and those that a name “would be good to have”.

Reviewer: ```suggestion
We start by dividing between two types of properties for content names, that is, those that a name _must_ have and those that a name _should_ have.
```

Give the audience, we might as well use RFC 2119 terminology :)

> In a content-oriented architecture (such as IPFS) the name **has to:**

Reviewer: ```suggestion
In a content-oriented architecture (such as IPFS) the name **must**:
```

- **provide security** guarantees, which further splits into two sub-properties
  - be *persistent & unique:* the name needs to stay the same even if the storage location or the owner of some content changes. Unique names enable the very important property of being able to cache content in arbitrary points in the network. Even if a tiny portion of the content changes, the name identifier changes too. In IPFS, this is achieved with hashing. Name-persistency makes it difficult to address dynamic content - see discussion on *“mutable content”* further below.
  - be *self-certifying:* the main purpose of having self-certifying names is to prove _data authenticity_. This is an important complementary part of name persistency to enable location-agnostic caching. Self-certifying names are normally non-human readable. 

> Traditionally, self-certifying names have been consider to be those that are not human readable and most commonly have been based on the hash of the content itself. Alternative ways to authenticate content is through obtaining the hash of the owner’s key. This helps in case of dynamic content and is used in IPNS and NDN, but it might lead to limited decentralisation properties.

Reviewer: ```suggestion
Traditionally, self-certifying names have been considered to be those that are not human-readable and most commonly have been based on the hash of the content itself. An alternative way to authenticate content is through the hash of the owner’s key. This helps in case of dynamic content and is used in IPNS and NDN, but it might lead to limited decentralisation properties.
```

> The second set of properties are those that **would be good to have** or need to be supported due to other core infrastructure requirements. These are:

Reviewer: ```suggestion
The second set of properties are those that the name **should** have or that need to be supported due to other core infrastructure requirements. These are:
```

Multibase now has support for a [mnemonic base](https://github.com/multiformats/js-multibase/compare/36d60d3f9379ea005327fe3a375f03ba286d0ecc...tableflip:feat/mnemonic-base) and given that CIDs are encoded with multibase so that they can be represented as strings, you can get your own mnemonic base encoded content address. This makes a name pronunceable, but given the amount of information in a CID, it is a long mnemonic.

##### IPFS link-shortening service provided by textile.photos cafes

Full explanation at https://blog.textile.io/ipfs-experiments-creating-ipfs-links-that-you-can-delete/

> 

Author: @yiannisbot is this and ask or a suggestion for later?

##### [DNS](https://en.wikipedia.org/wiki/Domain_Name_System)

One of the first widely deployed solutions for Human readable names.

##### ICN/NDN

In ICN/NDN names are used to do routing in the network (i.e., involving network routers). The NDN (formerly CCNx) project started with hierarchical, human-readable name structure, but it later became apparent that having only human-readable names is not quite an option.

It's impossible to guarantee that no two users will give the same name to some content. This translates to eventually needing to introduce a (logically) centralised entity (DNS-like) to keep track of these. Furthermore, most such ICN architectures assume existence of ASes and ISPs, as well as some entity similar to ICANN.

You can learn more about this at the following references:

- [Secure Naming for a Network of Information](https://core.ac.uk/download/pdf/11433344.pdf)
- [A Survey of Naming and Routing in Information-Centric Networks](https://www.researchgate.net/publication/260670278_A_Survey_of_Naming_and_Routing_in_Information-Centric_Networks)
- [Where is in a Name: A Survey of Mobility in ICN](http://www.eecs.qmul.ac.uk/~tysong/files/CACM13.pdf)

> 

Author: @yiannisbot, wanna add these in?
